Concurrency fixes in status.Manager

timstclair opened the pull request:
New & improved! Now with more fixes!
- Fix deadlock when syncing deleted pods with full update channel (successful termination case in unit test)
- Prevent sending stale updates to API server
- Don't delete cached status when sync fails (causes problems for prober)
- Fixes https://github.com/kubernetes/kubernetes/issues/16195

cc/ @kubernetes/goog-node

tallclair committed:
Concurrency fixes in status.Manager

- Fix deadlock when syncing deleted pods with full update channel
- Prevent sending stale updates to API server
- Don't delete cached status when sync fails (causes problems for prober)

timstclair commented on pkg/kubelet/status/manager.go:
```diff
@@ -182,16 +186,37 @@ func (m *manager) TerminatePods(pods []*api.Pod) bool {
-		select {
-		case m.podStatusChannel <- podStatusSyncRequest{pod, pod.Status}:
-		default:
+		newStatus := m.updateStatusInternal(pod, pod.Status)
```
Note to reviewers: This is a slight change in behavior, as we weren't caching the termination status before. I think this is the more correct way, although it does not enforce the same invariants as `SetPodStatus` (maintain `StartTime` and `LastTransitionTime`) - but I don't think that is necessary here.

smarterclayton replied:
That should be fine - my primary concern originally was not introducing more change than necessary in status, and fixing that is definitely desirable.

timstclair commented:
It just occurred to me that there's a risk (very low) of the server being update starved if versions change more rapidly than updates are processed. Let's discuss on Monday.

yujuhong commented:
Thanks for the nice PR! :-)

> It just occurred to me that there's a risk (very low) of the server being update starved if versions change more rapidly than updates are processed. Let's discuss on Monday.

The chance is low, but it could be extremely annoying if that happens. I think removing the `isStale()` check in `syncBatch` could solve this at the cost of sending more updates, which is anyway what we do today.

Not directly related to this, but I filed an issue a few days back to handle the `DeletePodStatus` situation differently (#16195). In this PR, we still rely on the fact that status is "populated" periodically by either prober or the pod worker (since the entries are marked stale on error). I would like to remove this assumption to free the pod worker from redundant periodic syncs. Some initial ideas to achieve this: 1) set the dirty bit in the cache and trigger a flush periodically if needed, along with the original channel or 2) remove the channel and simply use a work queue model in which the goroutine wakes up and check whatever needs to be sync'd. WDYT?

timstclair commented:
Made change to regularly sync "out of date" statuses to the API server, and track api server versions separately, so that update order is enforced without the potential to update-starve the API server.

timstclair commented:
tests look like flakes

yujuhong commented on pkg/kubelet/status/manager.go:
```diff
@@ -182,16 +195,42 @@ func (m *manager) TerminatePods(pods []*api.Pod) bool {
+		newStatus := m.updateStatusInternal(pod, pod.Status)
+		if newStatus != nil {
+			select {
+			case m.podStatusChannel <- podStatusSyncRequest{pod.UID, *newStatus}:
```
Do we still need to hold the lock while sending the request to the channel? I think we no longer have to do this, similar to what SetPodStatus does now.

timstclair replied:
We can do what SetPodStatus does, although now that we have the periodic sync, I actually wonder which approach is better. Perhaps it would be better to drop the update in the full channel case in both places and let the periodic sync pick up the update? Perhaps I'm over optimizing here.

yujuhong replied:
I like the idea of letting periodic sync dealing with this, but I am afraid that it may increase the latency -- user deletes a terminated pod has to wait the extra 0~10s for the pod to actually be deleted.

timstclair replied:
Only in the case where the channel is full - I'm not proposing removing the channel altogether, just falling back to the periodic sync when the channel is full.

yujuhong replied:
Ah I see. Sounds good. The "channel is full case" currently relies on our periodic housekeeping in kubelet, but would be good to simply fall back to the periodic sync.

yujuhong commented on pkg/kubelet/status/manager.go:
```diff
@@ -211,56 +250,75 @@ func (m *manager) RemoveOrphanedStatuses(podUIDs map[types.UID]bool) {
+	glog.Warningf("Failed to updated status for pod %q: %v", kubeletutil.FormatPodName(pod), err)
+}
+
+// needsUpdate returns whether the status is stale for the given pod UID.
```
Is this method thread-safe? Given that it's accessing the apiStatusVersions map, I'd assume it's not thread-safe. I guess only one goroutine can call this function, so we don't care?

timstclair replied:
It's not thread safe, but doesn't need to be. I added a couple of comments to that effect.

yujuhong replied:
SG. just wanted to confirm :)

yujuhong commented on pkg/kubelet/status/manager.go:
```diff
@@ -211,56 +250,75 @@ func (m *manager) RemoveOrphanedStatuses(podUIDs map[types.UID]bool) {
+	}
+}
+
+// syncPod syncs the given status with the API server. The caller must not hold the lock.
```
In line 266, the caller holds the podStatusesLock...

timstclair replied:
Oo... that's embarrassing, thanks for catching this! I went back and forth on the design here, and apparently left it in a half-way state. Fixed caller & fixed deadlock test to actually catch this.

yujuhong commented on pkg/kubelet/status/manager.go:
```diff
@@ -211,56 +250,75 @@ func (m *manager) RemoveOrphanedStatuses(podUIDs map[types.UID]bool) {
+		pod.Status = status.status
 		// TODO: handle conflict as a retry, make that easier too.
-		statusPod, err = m.kubeClient.Pods(pod.Namespace).UpdateStatus(statusPod)
+		pod, err = m.kubeClient.Pods(pod.Namespace).UpdateStatus(pod)
```
I think we need to delete the status and cached apiserver version if pod is not found here. Otherwise, we'd have leaked memory if someone else (e.g. a garbage collector) deletes the pod from etcd.

timstclair replied:
I'm a little confused by the positioning of this comment - where is "here"? Do you mean above where we handle the NotFound error? And maybe also where we handle the recreated case? We have cleanup loops for both orphaned statuses and apiStatusVersions, but I agree that we should try to minimize the need for cleaning up there.

yujuhong replied:
Yes, I meant handling the NotFound error here, but you're right about the cleanup loops. There should be no need to clean up immediately. In that case, perhaps we should simply remove the `DeletePodStatus`method completely. The only use case is line 307, which is already covered by the `RemoveOrphanedStatuses`. WDYT?

timstclair replied:
My inclination is to cleanup ASAP and only use the periodic cleanup loops as backup, because it doesn't keep stale data lying around and is more clear that it is actually getting cleaned up.

timstclair replied:
Discussed offline. Use cleanup loop for apiStatusVersion cleanup since it's internal only. Delete cached podStatus ASAP since other parts of the system (e.g. probe workers) rely on it.

yujuhong replied:
> Delete cached podStatus ASAP since other parts of the system (e.g. probe workers) rely on it.

Just wanted to clarify: not deleting right away should not cause any problem, so this is simply a slight performance optimization. In some cases it wouldn't help, e.g., if the watch is late to receive the pod removal, pod worker may still try to create containers and posts statuses.

yujuhong commented on pkg/kubelet/status/manager.go:
```diff
@@ -211,56 +250,75 @@ func (m *manager) RemoveOrphanedStatuses(podUIDs map[types.UID]bool) {
-				glog.V(3).Infof("Pod %q fully terminated and removed from etcd", statusPod.Name)
+			if err := m.kubeClient.Pods(pod.Namespace).Delete(pod.Name, api.NewDeleteOptions(0)); err == nil {
+				glog.V(3).Infof("Pod %q fully terminated and removed from etcd", pod.Name)
 				m.DeletePodStatus(pod.UID)
```
perhaps we can unexport `DeletePodStatus` since no one else calls it. It doesn't have to be in this PR though.

timstclair replied:
Done.

timstclair commented:
Pushed update from feedback. I had one question wrt deleting pod status (I think I addressed it?), also please take another look at SetPodStatus since I updated it with the fallback (like TerminatePods) as discussed above.

yujuhong commented on pkg/kubelet/status/manager.go:
```diff
@@ -211,56 +250,75 @@ func (m *manager) RemoveOrphanedStatuses(podUIDs map[types.UID]bool) {
+		if len(pod.UID) > 0 && pod.UID != uid {
+			glog.V(3).Infof("Pod %q was deleted and then recreated, skipping status update",
+				kubeletutil.FormatPodName(pod))
+			delete(m.apiStatusVersions, uid)
```
Do we need to delete apiStatusVersions here if we don't even delete the status?

yujuhong commented:
Thanks for the fix and extensive cleanup!

timstclair commented:
@k8s-bot unit test this

timstclair commented:
Squashed.

timstclair commented:
@k8s-bot unit test this
